### ISSUE TITLE
Add DELETE endpoint

### ISSUE THREAD

**Issue by kcibul** [broadinstitute/cromwell#1292]
As a user who runs cromwell in a production setting (like @ktibbett), I need to be able to manage the lifecycle of workflows in the system.  After running many workflows, they consume a lot of space on disk and even within the cromwell environment.  I woul to be able to delete them through a REST endpoint.

Add a new endpoint at DELETE/workflows/{version}/{id} which effectively removes this workflow from the system.  This should include

- removing all output files for the workflows and calls
- removing all metadata from the metadata service
- removing all workflows/calls from the call caching service

attempting to remove a workflow in a non-terminal state should result in an error (it should either finish or be aborted first).

**Comment by kshakir** [comment]
Further discussed points:

- The delete endpoint will be implemented by immediately marking a workflow as effectively "to be deleted"
- An actor will occasionally sweep through these "to be deleted" workflows and remove the previously discussed bullet points
- Upon success, the workflow will be gone from the database
- Upon failure, the workflow will be left in a "failed to delete" state in the database
- One may re-mark an existing workflow as "to be deleted" by re-using the delete endpoint
- If a workflow output is already deleted, we won't try to re-delete it. However--
- As the refresh token will not be available, output files that were generated with refresh tokens will _not_ be able to be deleted. These types of errors, while expected, will require more planning on how to handle these permissions issues.

**Comment by kcibul** [comment]
I think we need to talk about this more.  It sounds like the approach won't
work for FireCloud which is probably one of the most important use cases.



-------------------------------
Kristian Cibulskis
Chief Architect, Data Sciences & Data Engineering
Broad Institute of MIT and Harvard
kcibul@broadinstitute.org

**Comment by kcibul** [comment]
I think we need to talk about this more.  It sounds like the approach won't work for FireCloud which is a very important use case.  This might be a bigger feature that needs more refinement.

**Comment by geoffjentry** [comment]
@kcibul Considering there was some necromancy on this issue and some things were lost to the sands of time, is there any chance you remember why FC couldn't deal with this?

**Comment by kcibul** [comment]
It was just a feature that all users would want, and all would have to build themselves.  Once you run a workflow, if you want to get rid of it (esp on object storage) you either need to

a.) trawl the buckets with an implied understanding of the path structure of Cromwell, then delete those files

b.) use cromwell metadata to identify all the files (which is sort of a PITA, Banks is doing that now) and then delete those.

So... client's _could_ do this, but it's just boilerplate code (and annoying)

**Comment by geoffjentry** [comment]
@kcibul I was talking about this bit you said above: "It sounds like the approach won't work for FireCloud" - I agree it makes sense to do this in general :)

**Comment by katevoss** [comment]
This issue deserves it's own spec document so I'm giving it one: [The Delete Spec](https://docs.google.com/document/d/1ShP0_a2nUFWr8lh0xISPtrpBpp5og-Hkdv-bWoB4tMY/edit).

Feel free to opine there :)

**Comment by katevoss** [comment]
As a **workflow runner**, I want **to delete information I no longer need from a workflow**, so that I can **stop paying for storage for it**.
- Effort: **Large**
- Risk: **Medium to Large**
- - implications for call caching?
- - provenance?
- Business value: **Medium to Large**
- - big cost-saver

**Comment by kcibul** [comment]
This one comes up *a lot*, and will even more once people can see how much they are spending on intermediate results in FireCloud

I wanted to ask about the effort -- why large?  Alex has already implemented the logic for what to delete by looking at the workflow metadata, and if we simply had an endpoint to "evict from call cache" we could build this outside the system pretty easily.  Basically, iterate through a workflow and for every "output" that is not part of the workflow "output", you rm the file and evict the call from the task that produced it (or maybe any task that uses that output?).  What are you thinking of that I'm missing?  Or maybe I'm interpreting "Large" as being bigger than you are?

**Comment by geoffjentry** [comment]
I have a fear that *Large* came from me, but I don't know why I would have said that. Thinking about it right now the only part which seems tricky is making sure people understand the implications for call caching

**Comment by geoffjentry** [comment]
@kcibul Re-reading the issue, one thing is *not* going to be in the cards. We will *not* be removing the workflow from the metadata. The point of metadata is that it is immutable (our unfortunate fondness for migrations notwithstanding). What we *can* do is to send an event to metadata saying that the workflow has been deleted.

**Comment by kcibul** [comment]
So for intention of this endpoint (storage savings) that's probably fine, although I would highly recommend we rename this endpoint to be "compact" or something.

At some point, it would be good to have a full data lifecycle delete or provide ways to manage the "db storage" behind Cromwell without having people blindly monkey around in the db.  scripts or docs could accomplish this too

**Comment by geoffjentry** [comment]
It just goes back to the point of the metadata architecture. It's entire intention is to be write once and then read only. People should also not be blindly removing things from it either. Totally agree w/ your 2nd paragraph but the "remove stuff from metadata" option is a pretty fundamental change - there are other ways to attack it if it's an issue.

**Comment by kcibul** [comment]
@katevoss In re-reading this thread...  I think the MVP has to include the ability to delete things via refresh token somehow.  (Which looks like I said the same thing on this issue almost a year ago!)

**Comment by geoffjentry** [comment]
@kcibul Could the refresh token aspect be what caused you to say the following: " It sounds like the approach won't work for FireCloud which is a very important use case."? I've asked a few times in this thread what that meant and am still in the dark :)

**Comment by kcibul** [comment]
Yes -- my original comment was because of the inability to remove files created via refresh token.  I think then you asked why FireCloud couldn't implement the delete functionality themselves... which they could but that seems brittle and better to centralize (e.g. they would have to crawl the metadata, evict things from the cache, etc).

**Comment by geoffjentry** [comment]
I definitely don't think they should be doing it themselves for those very reasons so rest easy :) at least not in the long term, there's discussion over there about doing something in the interim

**Comment by eitanbanks** [comment]
FWIW I want this functionality!  With 20K shards every time we create a joint callset, that's a ton of files to be listing (and then paying for that list command)...

**Comment by geoffjentry** [comment]
Was talking with faces w/ @abaumann about this, particularly as it regards to current problems Firecloud is having. We managed to seriously birdstone a variety of issues by adding the following tidbit to this feature request:

- The deletion of files is optional. If they choose to not delete their files, Cromwell will just erase traces of the workflow from its database

**Comment by eitanbanks** [comment]
Thinking about this a little more, I just added this comment to Kate's doc:
"FWIW, I don’t necessarily need Cromwell to do the delete for me in my use case.  I’m happy to do it if I have all the files at hand that I want to delete.  This is important because I don’t really want an all-or-nothing delete.  I usually want to delete the call cached intermediate results, but not the final outputs of the workflow (and I note that not all pipeline developers list the outputs in the wdl -- I just learned this yesterday from Megan -- so I would like some control over the delete in many cases)."

**Comment by katevoss** [comment]
After much discussion and clarification (thanks @geoffjentry ), these are the 2 use cases relating to Deleting and Cleaning up files after running a workflow:
* Use case 1: I want to clean up intermediary files from my workflow, but keep all outputs. I understand that I will no longer be able to call cache on this workflow.
* Use case 2: I want to delete all files related to my workflow, I will never need it again. This includes outputs. I will not be able to call cache on this workflow.

There are a few remaining questions to answer but we are getting closer. Thank you all for your input in the meantime.

**Comment by geoffjentry** [comment]
There really are 3.

- Remove from Cromwell but don't touch any files (for Firecloud)
- Remove from Cromwell and delete all files
- Remove from Cromwell and delete intermediate outputs

**Comment by ernoc** [comment]
I'd like to suggest resuming focus on this feature request. I believe this is much needed, as not being able to delete (especially intermediate) files from old workflows may become a stopper from using cromwell on production scale. Intermediate files together could be much larger than final result files, increasing storage requirements by orders of magnitude. Many thanks!

**Comment by ruchim** [comment]
@ernoc We hope to have this feature released in the next few months that involves being able to cleanup workflow outputs!

**Comment by ruchim** [comment]
AC: 
1. Delete endpoint that targets deleting intermediate output files.
2. Cleanup of those intermediate outputs if they were utilized by the cache store.

**Comment by danbills** [comment]
Detailed writeup here: https://docs.google.com/document/d/1aJn5HzvDgYbvBlEG4z0KO8oZgaQ3lFu2hE8QzRC0_18/edit?usp=sharing

**Comment by caross73** [comment]
I want to add that we need this functionality or something similar. We are running cromwell workflows as a production service and are in the awkward position of  putting cromwell into standalone mode periodically scrub so that we can flush and rebuild the backing database to conserve disk space.
